### DISCUSSION TITLE
Head/Face Pose orientation (Yaw, Pitch, Roll) estimate from the FaceMesh Model.

### DISCUSSION THREAD

**Discussion by saeidn95** [vladmandic/human#433] [Q&A]
I need some help to obtain the head/face Pose orientation (Yaw, Pitch, Roll) estimate from the MediaPipe FaceMesh Model. Can anyone help? I would also welcome info about any other models on your repo that give me the (Yaw, Pitch, Roll) info for the head/face.

**Comment by vladmandic** [comment]
is the question how to calculate yaw/pitch/roll if you're using facemesh directly without `human`?
because if you're using human, its there in results. and if you're not, feel free to look through source codebase, starting with <https://github.com/vladmandic/human/blob/main/src/face/angles.ts>

**Reply by saeidn95** [reply]
Thanks for the quick response! I was looking follow the codebase in human/src/face as guide, and to build my own code in C++, using ONNX runtime API. I think angles.ts is what I am looking for. I assume angles.ts is a post processing step after the faceMesh model inference.

**Reply by saeidn95** [reply]
I have one related question about the FaceMesh. The guid for MeshFace in MediaPipe (https://github.com/google/mediapipe/blob/master/docs/solutions/face_mesh.md), refers to an static 3D "Canonical Face Model" Where in human you deal with this or is all internal to the faceMesh model?

**Reply by vladmandic** [reply]
why would that be needed? we're either talking about point-cloud or a projected uvmap for 3d mesh.
canonical face model is just noted on how it works - facemesh does not detect points as such as most models, but instead performs transforms on the canonical face thus moving the points of the face to fit whats detected.

**Reply by saeidn95** [reply]
I don't think I would need it as such for my code; but trying to go a bit into the weeds to see how it all works. and take a closer look as how the transforms on the canonical face work in the lines of code. I assume it is all embedded in the faceMesh model.

**Reply by saeidn95** [reply]
"starting with https://github.com/vladmandic/human/blob/main/src/face/angles.ts" This was really useful. Thank you very much! A couple of questions a) what is the typical accuracy of (yaw, pitch roll, and eye gaze) b) what is the gaze bearing is with respect to? Is it with  respect to roll or something else? and what is the gaze strength signify?

**Reply by vladmandic** [reply]
yaw/pitch/roll are decent accuracy for small-mid angles, but it drops drastically for large angles
gaze is calculated purely based on retina position relative to eye bounding box, so they are heavily dependent on retina being clearly visible (so if you look down for example, eye lids are covering retina and gaze will not be accurate)

**Reply by saeidn95** [reply]
"gaze is calculated purely based on retina position relative to eye bounding box" Shouldn't we have two angles for the gaze then? One with respect to up/bottom side and one with respect to left/right side?

**Reply by vladmandic** [reply]
"up" is prone to errors, your eyelid is almost never wide open

**Reply by saeidn95** [reply]
I understand now! What about strength part of the gaze? Does signify the size of the gaze arrow proportional to the eye opening?

**Reply by vladmandic** [reply]
if you to the look left just a little or all the way, arrow direction is the same - length of arrow is how much left, meaning offset of retina from eye cemter.

**Reply by saeidn95** [reply]
Great! thank you very much. BTW do you have a python based version of this code base?

**Reply by vladmandic** [reply]
no, this is js only for now.
